src/rbac/default_role_manager.rs

```diff
@@ -0,0 +1,353 @@
+use std::cell::RefCell;
+use std::rc::Rc;
+use std::collections::HashMap;
+use crate::rbac::RoleManager;
+
+type MatchingFunc = fn(&str, &str) -> bool;
+
+#[derive(Clone)]
+pub struct DefaultRoleManager {
+    pub all_roles: HashMap<String, Rc<RefCell<Role>>>,
+    pub max_hierarchy_level: usize,
+    pub has_pattern: bool,
+    pub matching_func: Option<MatchingFunc>,
+}
+
+impl DefaultRoleManager {
+    pub fn new(max_hierarchy_level: usize) -> Self {
+        return DefaultRoleManager {
+            all_roles: HashMap::new(),
+            max_hierarchy_level,
+            has_pattern: false,
+            matching_func: None,
+        };
+    }
+
+    fn create_role(&mut self, name: &str) -> Rc<RefCell<Role>> {
+        self.all_roles
+            .entry(name.to_owned())
+            .or_insert(Rc::new(RefCell::new(Role::new(name.to_owned()))))
+            .clone()
+    }
+
+    fn has_role(&self, name: &str) -> bool {
+        if let Some(_role) = self.all_roles.get(name) {
+            return true;
+        }
+        return false;
+    }
+}
+
+impl RoleManager for DefaultRoleManager {
+    fn clone_box(&self) -> Box<dyn RoleManager> {
+        Box::new(self.clone())
+    }
+
+    fn add_link(&mut self, name1: &str, name2: &str, domain: Vec<&str>) {
+        let mut name1 = name1.to_owned();
+        let mut name2 = name2.to_owned();
```

Comment:
It appears unnecessary to create owned references to these strings.

Reply:
I'm using `to_owned` here to convert `&str` to `String`, because `format!` need `String` as parameters. Do you suggest using `concat!` here?